src/labone/dataserver.py

```diff
@@ -0,0 +1,165 @@
+            hide_kernel_prefix: Hide the kernel prefix in the nodetree.
+                Default is `True`.
+            parser_builder: A custom parser builder for the nodetree.
+                Default is `None`.
+
+        Returns:
+            The connected device.
+        """
+        session = await KernelSession.create(
+            kernel_info=DeviceKernelInfo(device_id=serial, interface=interface),
+            server_info=ServerInfo(host=self.host, port=self.port),
+        )
+
+        try:
+            model_node = await construct_nodetree(
+                session,
+                hide_kernel_prefix=hide_kernel_prefix,
+                use_enum_parser=use_enum_parser,
+                custom_parser=custom_parser,
+            )
+        except LabOneError as e:
+            raise LabOneError(
+                f"Could not connect to device {serial} through {interface}.",
+            ) from e
+
+        return Instrument(
+            serial=serial,
+            data_server=self,
+            model_node=model_node,
+        )
+
+    @property
+    def host(self) -> str:
+        """Host of the Data Server."""
+        return self._host
+
+    @property
+    def port(self) -> int:
+        """Port of the Data Server."""
+        return self._port
+
+    async def _check_firmware_update_status(self) -> None:
+        """Check if the firmware matches LabOne version.
+
+        Raises:
+            ConnectionError: If the device is currently updating
+            LabOneError: If the firmware revision does not match to the
+                version of the connected LabOne DataServer.
+        """
+        devices_ = await self.tree_manager.session.get('/zi/devices')
+        devices: dict[str, NodeInfo] = json.loads(
+            (devices_).value  # type: ignore
+        )
+        for device_id, device_info in devices.items():
+            status_flag = device_info["STATUSFLAGS"]
+            if status_flag & 1 << 8:
+                raise ConnectionError(
+                    "The device is currently updating please try again after the update "
+                    "process is complete",
+                )
+            if status_flag & 1 << 4 or status_flag & 1 << 5:
+                raise LabOneError(
+                    "The Firmware does not match the LabOne version. "
+                    "Please update the firmware (e.g. in the LabOne UI)",
+                )
+            if status_flag & 1 << 6 or status_flag & 1 << 7:
+                raise LabOneError(
+                    "The Firmware does not match the LabOne version. "
+                    "Please update LabOne to the latest version from "
+                    "https://www.zhinst.com/support/download-center.",
+                )
+
+    async def check_compatibility(self) -> None:
+        """Check if the firmware revision matches the LabOne Data Server version.
+
+        Raises:
+        """
+        await self._check_firmware_update_status()
```

Comment:
I would make this a single function. And name it `check_firmware_compatibility`

Also I would change the logic in a way that it iterates through all devices and checks for mismatch/update. And than at the end raise an error which in the msg has all found mismatches ... otherwise a user only sees a single error and needs to have multiple iterations

Reply:
This is still not fixed